### PR TITLE
Feat: pivot role limit kms

### PR DESCRIPTION
**Pull request by dlpzx**
### Feature or Bugfix
- Feature

### Detail
- read KMS keys with an alias prefixed by the environment resource prefix
- read KMS keys imported in imported datasets
- restrict pivot role policies to the KMS keys created by data.all and those imported in the imported datasets
- move kms client from data_sharing to base as it is used in environments and datasets

### Relates
- #580

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

This PR restricts the IAM policies of the pivot role, following the least privilege permissions principle

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by dlpzx** [comment]
I have tested locally with a real imported datasets and some fake datasets to put the iam_policy util under stress.
For 500 imported datasets these are the resulting pivot role policies:

Screenshot 2023-10-26 at 14 09 57


Screenshot 2023-10-26 at 14 10 19

File: backend/dataall/base/aws/kms.py [hunk 1]
```diff
@@ -69,3 +69,17 @@ def check_key_exists(self, key_alias: str):
             return None
         else:
             return key_exist
+
```
Comment (dlpzx) [inline backend/dataall/base/aws/kms.py:72]: This function is not used at the moment. @noah-paige should we remove it and move the `kms` handler back to the data_sharing module?

Comment (dlpzx) [reply backend/dataall/base/aws/kms.py]: I moved it back to avoid issues with the work done in #284

**Comment by noah-paige** [comment]
Tested this Pr out in an AWS Deployment - all permissions added to PivotRole look correct but I am also facing the same error: 
`An error occurred (MalformedPolicyDocumentException) when calling the PutKeyPolicy operation: The new key policy will not allow you to update the key policy in the future.`

When doing folder sharing on imported buckets only

**Comment by noah-paige** [comment]
> Tested this PR out in an AWS Deployment - all permissions added to PivotRole look correct but I am also facing the same error: `An error occurred (MalformedPolicyDocumentException) when calling the PutKeyPolicy operation: The new key policy will not allow you to update the key policy in the future.`
> 
> When doing folder sharing on imported buckets only

After a bit of testing - I found the only way I could resolve the `MalformedPolicyDocumentException` is to explicitly add in the imported KMS Key Policy the following permission set to the PivotRole:
```
                "kms:Decrypt",
                "kms:Encrypt",
                "kms:GenerateDataKey*",
                "kms:PutKeyPolicy",
                "kms:GetKeyPolicy",
                "kms:ReEncrypt*",
                "kms:TagResource",
                "kms:UntagResource"
```

Even if you enabled IAM privileges to the KMS key policy similar to:
```
{
  "Sid": "Enable IAM User Permissions",
  "Effect": "Allow",
  "Principal": {
    "AWS": "arn:aws:iam::111122223333:root"
   },
  "Action": "kms:*",
  "Resource": "*"
}
```

and then add the same kms permissions set to the Pivot Role's IAM Policies it would not work (needed to be explicitly in KMS)

This could be helpful to call out in documentation but implementation would be required by the data.all user importing the key to ensure that they are giving proper access to data.all's pivotRole for their pre-created infra

**Comment by dlpzx** [comment]
@noah-paige thanks for the review and for looking into the `MalformedPolicyDocumentException` I think we can tackle that issue with documentation, I added a small comment in the import section

File: documentation/userguide/docs/datasets.md [hunk 2]
```diff
@@ -101,6 +101,42 @@ the fields of a newly created dataset you have to specify the S3 bucket and opti
 is left empty, data.all will create a Glue database pointing at the S3 Bucket. As for the KMS key Alias, data.all assumes that if nothing is specified
 the S3 Bucket is encrypted with SSE-S3 encryption.
 
+!!! danger "Imported KMS key and S3 Bucket policies requirements"
+    Data.all pivot role will handle data sharing on the imported Bucket and KMS key (if imported). Make sure that
+    the resource policies allow the pivot role to manage them. For the KMS key policy, explicit permissions are needed. See an example below.
+
+
+### KMS key policy
+In the KMS key policy we need to grant explicit permission to the pivot role. Note that this block is needed even if
```
Comment (zsaltys) [inline documentation/userguide/docs/datasets.md:110]: @dlpzx this is not correct. You do not need to do this if the key has root level policy.. then the permissions on pivotRole are enough

**Pull request merged**